### PR TITLE
HHH-17643 Allow uninitialized proxy serialization even when a SessionFactory is not available

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-17643

Alternative solution to #7700.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java [hunk 1]
```diff
@@ -32,8 +49,33 @@ public final class BytecodeProviderInitiator implements StandardServiceInitiator
...
+			return new org.hibernate.bytecode.internal.none.BytecodeProviderImpl();
+		}
+
+		BytecodeProvider provider = null;
+		while ( iterator.hasNext() ) {
+			final BytecodeProvider next = iterator.next();
+			if ( provider == null ) {
+				provider = next;
+			}
+			else if ( provider instanceof org.hibernate.bytecode.internal.none.BytecodeProviderImpl ) {
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java:70]: I'm a little confused about the purpose of these checks, could you elaborate please?

Wasn't it simpler to pass the `bytecodeProviders` parameter as a Collection, and check its size?
if:
 0 : use the no-op
 1 : use whatever was found
 >1 : throw an error

File: hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java [hunk 2]
```diff
@@ -43,12 +85,15 @@ public Class<BytecodeProvider> getServiceInitiated() {
...
+		return getBytecodeProvider( ServiceLoader.load( BytecodeProvider.class ) );
 	}
 
+	/**
+	 * @deprecated Register a {@link BytecodeProvider} through Java {@linkplain java.util.ServiceLoader services}.
+	 */
 	@Internal
+	@Deprecated( forRemoval = true )
 	public static BytecodeProvider buildBytecodeProvider(String providerName) {
-
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java:51]: I like that you're being conservative, but considering it's marked as internal it might be ok to just remove this method, or perhaps replace its body with a "throw error".
My reasoning is that we should get rid of those references to the `org.hibernate.bytecode.internal.bytebuddy.BytecodeProviderImpl` as soon as possible..

But I agree perhaps do it in a separate commit, so to change as least as possible for the patch to be backported, for example I'm not sure what impact it would have on Spring native as they also have some needs in this area.

File: hibernate-core/src/main/java/org/hibernate/cfg/BytecodeSettings.java [hunk 3]
```diff
@@ -20,7 +20,10 @@ public interface BytecodeSettings {
 	 * At present only bytebuddy is supported, bytebuddy being the default since version 5.3.
 	 *
 	 * @settingDefault {@code "bytebuddy"}
+	 * @deprecated Will be removed, users can override the default provider by registering a
+	 * service for the BytecodeProvider interface. See HHH-17643
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/cfg/BytecodeSettings.java:24]: Great that you remembered to update the documentation :+1: 

However I'd change the language a little: people don't "register"  anything to t he service loader, but things will be found automatically as long as they have the right dependency on the classpath.

So it might be more useful to explain something among those lines: "Hibernate ORM will use the BytecodeProvider implementation it finds on the classpath loading it via the standard ServiceLoader mechanism. Currently there is only a single implementation, which is included in Hibernate ORM so it's currently not possible to override this."

File: hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java [hunk 4]
```diff
@@ -30,6 +35,8 @@ public final class SerializableProxy extends AbstractSerializableProxy {
 
 	private final CompositeType componentIdType;
 
+	private static volatile BytecodeProviderImpl defaultProvider;
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java:38]: This is a bit tricky: if `defaultProvider` gets initialized (for wathever strange reason) during the before-bootstrap phase in a native image compilation, it will get stuck on the wrong reference.

Some options:

 A) Store it somewhere else in a non-static field (seems hard to find a sane place?)
 B) Not store it at all. Probably not good for performance but I'm not sure of how bad that would be, might not be too bad (?)
 C) Ensure this particular field gets reset by the GraalVM compiler; this could be done either via the annotation https://docs.oracle.com/en/graalvm/jdk/17/sdk/com/oracle/svm/core/annotate/RecomputeFieldValue.Kind.html ; we probably don't want the project to depend on such annotations but there should be an API one can use:
 - https://docs.oracle.com/en/graalvm/jdk/17/sdk/org/graalvm/nativeimage/hosted/Feature.BeforeAnalysisAccess.html#registerFieldValueTransformer-java.lang.reflect.Field-org.graalvm.nativeimage.hosted.FieldValueTransformer-  (let me know if you want to go down this path, in that case I can help..)

File: hibernate-core/src/main/resources/META-INF/services/org.hibernate.bytecode.spi.BytecodeProvider [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#
```
Comment (Sanne) [inline hibernate-core/src/main/resources/META-INF/services/org.hibernate.bytecode.spi.BytecodeProvider:1]: Do we need the copyright header in here?
Unless checkstyle is complaining, I'd keep it simple.

**Review by Sanne — CHANGES_REQUESTED** [review]
Great work, that's approximately what I'd have done.
There's one complexity with the static field cache... I've commented in there.

File: hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java (hunk 1 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java:70]: Isn't the `org.hibernate.bytecode.internal.none.BytecodeProviderImpl` always going to be provided by `hibernate-core` itself? I though if a user wanted to provide a custom implementation that would need to override the default, that's the reason of the checks.

I'm also passing an `Iterable<BytecodeProvider>` parameter to allow passing the result from `ServiceLoader.load( BytecodeProvider.class )` directly to this method.

File: hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java (hunk 4 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java:38]: Why would it be initialized, though? This method only gets called when trying to deserialize a `SerializableProxy` instance. Are there any reasons why this could happen during the "before-bootstrap" phase?

Apart from that, I'm not quite sure which of the options is the best one to follow to be honest, you can pick what you feel is the best way to go and I'll just update the PR with that.

File: hibernate-core/src/main/java/org/hibernate/cfg/BytecodeSettings.java (hunk 3 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/cfg/BytecodeSettings.java:24]: Changed the description with this, thanks.

File: hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java (hunk 2 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java:51]: Added a separate commit that removes the method.

File: hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java (hunk 4 above)
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/proxy/pojo/bytebuddy/SerializableProxy.java:38]: Yes I suppose you're right that the problem I raised is quite unlikely and difficult to trigger; it would require for some code to trigger serialization of some proxy before the classpath is changed - my concern stems from the possibility of `static` fields to be shared shared across multiple instances of Hibernate ORM, so if some SessionFactory(s) were to be initialized at a later stage, or perhaps rebooted, the cached field could witness a reuse of previous value - however I suppose in such a case we wouldn't also witness a change of the enumeration of service implementations available on the classpath, this would only be possible if the condition of multiple ORMs instances is combined with a native image compilation or some other runtime/container trick.

Feel free to ignore it for now - I admit it's not a practical problem as I don't think it can be triggered in practice in *today* in the frameworks we care about, I'll just prefer to avoid such possibilities as if it ever happens one day, perhaps because some of our assumptions about bootstrap are changed, it would be a very tricky one nail down.

In short, resetting the static field is the right thing to do in theory but I can take care of that later.

File: hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java (hunk 1 above)
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java]: > Isn't the org.hibernate.bytecode.internal.none.BytecodeProviderImpl always going to be provided by hibernate-core itself? I thought if a user wanted to provide a custom implementation that would need to override the default, that's the reason of the checks.

Ah yes I see what you mean now. I was assuming that we wouldn't allow to load a custom implementation for now, as I was expecting we'd first move the ByteBuddy implementation into an optional, separate jar.

But feel free to keep it then, it's nice that people won't have to wait for that additional refactoring step - I'm not expecting many users to want to do that though :)

**Review by Sanne — APPROVED** [review]
Nice, looks good to me.
Perhaps I'd change the name of `getDefaultProvider()` to either `getFallbackBytecodeProvider()` (as it's not the default but what we want to use when the SF is unavailable), or simply `getBytecodeProvider()` ? 

And same for the field name? Up to you :)

File: hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java (hunk 1 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java]: I've already simplified the code to just retrieve the first available implementation, and throw if there are more than one. I don't think anyone would be interested in providing custom implementations anyways, at least in the short term. We can always change this once we decouple the bytecode provider from hibernate-core.
